### DISCUSSION TITLE
How do I auto start the script on detached mode on reboot?

### DISCUSSION THREAD

**Discussion by SRoy89** [Tkd-Alex/Twitch-Channel-Points-Miner-v2#526] [Q&A] [answered]
I am running the script barebone on RaspberryPi 4, with this command `python /home/pi/Twitch-Channel-Points-Miner-v2/run.py`.
I don't much care about logs, altho some info is sent to discord via webhook.

So, how do I make it start automatically after reboots without having to keep the terminal window open all the time?
I tried to use the [example docker-composer](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2#docker), but the container didn't boot.

I also tried `nohup` but couldn't enter the code with this method when it asked for email verification.

Thanks.

**Comment by ColinShark** [comment, marked as answer]
You could set up a cronjob like so:
`@reboot python /home/pi/Twitch-Channel-Points-Miner-v2/run.py`
Keep in mind that this is non-standard and may or may not work.

Alternatively you could use something like [systemd](https://systemd.io/) or [supervisord](https://supervisord.org/) to have it auto-start. The former should be pre-installed but requires sudo, the latter can be run by any user.

Also regarding your email verification; you probably want to log in interactively first instead of extracting and copying your cookie. That way your miner can stay logged in in case you switch sessions on your desktop browser.

**Reply by SRoy89** [reply]
> Also regarding your email verification; you probably want to log in interactively first instead of extracting and copying your cookie. That way your miner can stay logged in in case you switch sessions on your desktop browser.

Do you mean first running it and logging in manually so it gets the cookies, then let it run automatically on boot? If not please explain what you mean by "interactively".

> Alternatively you could use something like [systemd](https://systemd.io/)

As you said cronjob might not work, so I used systemd. But it shows me error when I checked the status of the service after reboot.
I'm not sure what I did wrong here, I followed the [yt tutorial](https://youtu.be/DUGZC-tNm2w) exactly, but I'm willing to accept any help I can. Thanks!

Here's the `twitch-miner.service` code and the status error below it:
```python

[Unit]
Description=Twitch Points miner python script in home folder
After=multi-user.target

[Service]
ExecStart=/usr/bin/python /home/pi/Twitch-Channel-Points-Miner-v2/run.py
User=pi

[Install]
WantedBy=multi-user.target
```

```bash
● twitch-miner.service - Twitch Points miner python script in home folder
     Loaded: loaded (/lib/systemd/system/twitch-miner.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Wed 2022-06-15 15:08:05 IST; 33s ago
    Process: 2846 ExecStart=/usr/bin/python /home/pi/Twitch-Channel-Points-Miner-v2/run.py (code=exited, status=1/FAILURE)
   Main PID: 2846 (code=exited, status=1/FAILURE)
        CPU: 3.047s

Jun 15 15:08:05 raspberrypi python[2846]:   File "/home/pi/Twitch-Channel-Points-Miner-v2/TwitchChannelPointsMiner/TwitchChannelPointsMiner.py", line 83, in >
Jun 15 15:08:05 raspberrypi python[2846]:     Path(Settings.analytics_path).mkdir(parents=True, exist_ok=True)
Jun 15 15:08:05 raspberrypi python[2846]:   File "/usr/lib/python3.9/pathlib.py", line 1316, in mkdir
Jun 15 15:08:05 raspberrypi python[2846]:     self.parent.mkdir(parents=True, exist_ok=True)
Jun 15 15:08:05 raspberrypi python[2846]:   File "/usr/lib/python3.9/pathlib.py", line 1312, in mkdir
Jun 15 15:08:05 raspberrypi python[2846]:     self._accessor.mkdir(self, mode)
Jun 15 15:08:05 raspberrypi python[2846]: PermissionError: [Errno 13] Permission denied: '/analytics'
Jun 15 15:08:05 raspberrypi systemd[1]: twitch-miner.service: Main process exited, code=exited, status=1/FAILURE
Jun 15 15:08:05 raspberrypi systemd[1]: twitch-miner.service: Failed with result 'exit-code'.
Jun 15 15:08:05 raspberrypi systemd[1]: twitch-miner.service: Consumed 3.047s CPU time.
```

**Reply by ColinShark** [reply]
> Do you mean first running it and logging in manually so it gets the cookies, then let it run automatically on boot?

Yes, that's exactly what I meant by interactively :).
Run it yourself with `python run.py` so the program asks you for credentials, generating its own cookie.

---

For your service file you probably want to set [`WorkingDirectory=`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#WorkingDirectory=): _"If not set, defaults to [..] the respective user's home directory if run as user."_

```service
WorkingDirectory=/home/pi/Twitch-Channel-Points-Miner-v2/
```

That way your script will use relative paths from the program itself based on that working directory, instead of from where systemd was executed. Since that will most likely be `root` or `systemd`, your `pi` user doesn't have access to their `/analytics` path.

Take what I say with a grain of salt, I pretty much don't work with systemd, only supervisord 😅

**Reply by SRoy89** [reply]
That worked, thanks a bunch!
